subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -228,44 +300,112 @@ static void firmware_transfer(struct k_work *work)
 					       .sin_family = AF_INET };
 #endif
 	struct net_if *iface;
-	int ret, port, family;
+	int ret, family;
+	u16_t off;
+	u16_t len;
+	char tmp;
 
 	/* Server Peer IP information */
-	/* TODO: use parser on firmware_uri to determine IP version + port */
-	/* TODO: hard code IPv4 + port for now */
-	port = 5685;
 	family = AF_INET;
 
-#if defined(CONFIG_NET_IPV6)
-	if (family == AF_INET6) {
+	http_parser_url_init(&parsed_uri);
+	ret = http_parser_parse_url(firmware_uri,
+				    strlen(firmware_uri),
+				    0,
+				    &parsed_uri);
+	if (ret != 0) {
+		SYS_LOG_ERR("Invalid firmware URI: %s", firmware_uri);
+		lwm2m_firmware_set_update_result(RESULT_INVALID_URI);
+		return;
+	}
+
+	/* Check schema and only support coap for now */
+	if (!(parsed_uri.field_set & (1 << UF_SCHEMA))) {
+		SYS_LOG_ERR("No schema in package uri");
+		lwm2m_firmware_set_update_result(RESULT_INVALID_URI);
+		return;
+	}
+
+	/* TODO: enable coaps when DTLS is ready */
+	off = parsed_uri.field_data[UF_SCHEMA].off;
+	len = parsed_uri.field_data[UF_SCHEMA].len;
+	if (len != 4 || memcmp(firmware_uri + off, "coap", 4)) {
+		SYS_LOG_ERR("Unsupported schema");
+		lwm2m_firmware_set_update_result(RESULT_UNSUP_PROTO);
+		return;
+	}
+
+	if (!(parsed_uri.field_set & (1 << UF_PORT))) {
+		/* Set to default port of CoAP */
+		parsed_uri.port = 5683;
+	}
+
+	off = parsed_uri.field_data[UF_HOST].off;
+	len = parsed_uri.field_data[UF_HOST].len;
+
+	/* IPv6 is wrapped by brackets */
+	if (off > 0 && firmware_uri[off - 1] == '[') {
```

Comment:
BTW, I just submitted #1198 which provides net_ipaddr_parse() function that converts a IPv{4|6} address string to struct sockaddr. That function could probably be used here in order to save some memory, this is just something that could be considered here.